lib/Sema/SemaBounds.cpp

```diff
@@ -628,93 +698,210 @@ namespace {
           }
 
           // Pointer arithmetic.
+          //
+          // `p + i` has the bounds of `p`. `p` is an RValue.
+          // `p += i` has the lvalue target bounds of `p`. `p` is an LValue.
+          // same applies for `-` and `-=` respectively
           if (LHS->getType()->isPointerType() &&
               RHS->getType()->isIntegerType() &&
               BinaryOperator::isAdditiveOp(Op)) {
             return IsCompoundAssignment ?
               LValueTargetBounds(LHS) : RValueBounds(LHS);
           }
+          // `i + p` has the bounds of `p`. `p` is an RValue.
+          // `i += p` has the bounds of `p`. `p` is an RValue.
           if (LHS->getType()->isIntegerType() &&
               RHS->getType()->isPointerType() &&
-              BinaryOperator::isAdditiveOp(Op)) {
-            assert(!IsCompoundAssignment);
+              Op == BinaryOperatorKind::BO_Add) {
             return RValueBounds(RHS);
           }
+          // `e - p` has empty bounds, regardless of the bounds of p.
+          // `e -= p` has empty bounds, regardless of the bounds of p.
+          if (RHS->getType()->isPointerType() &&
+              Op == BinaryOperatorKind::BO_Sub) {
+            return CreateBoundsEmpty();
+          }
 
           // Arithmetic on integers with bounds.
+          //
+          // `e1 @ e2` has the bounds of whichever of `e1` or `e2` has bounds.
+          // if both `e1` and `e2` have bounds, then they must be equal.
+          // Both `e1` and `e2` are RValues
+          //
+          // `e1 @= e2` has the bounds of whichever of `e1` or `e2` has bounds.
+          // if both `e1` and `e2` have bounds, then they must be equal.
+          // `e1` is an LValue, its bounds are the lvalue target bounds.
+          // `e2` is an RValue
+          //
+          // @ can stand for: +, -, *, /, %, &, |, ^, >>, <<
           if (LHS->getType()->isIntegerType() &&
               RHS->getType()->isIntegerType() &&
               (BinaryOperator::isAdditiveOp(Op) ||
                BinaryOperator::isMultiplicativeOp(Op) ||
-               BinaryOperator::isBitwiseOp(Op))) {
+               BinaryOperator::isBitwiseOp(Op) ||
+               BinaryOperator::isShiftOp(Op))) {
             BoundsExpr *LHSBounds = IsCompoundAssignment ?
               LValueTargetBounds(LHS) : RValueBounds(LHS);
             BoundsExpr *RHSBounds = RValueBounds(RHS);
             if (LHSBounds->isNone() && !RHSBounds->isNone())
               return RHSBounds;
             if (!LHSBounds->isNone() && RHSBounds->isNone())
               return LHSBounds;
-            // TODO: check if both LHS and RHS have bounds and
-            // indicate cause of failure to have bounds.
+            if (!LHSBounds->isNone() && !RHSBounds->isNone()) {
+              // TODO: Check if LHSBounds and RHSBounds are equal.
+              // if so, return one of them. If not, return bounds(none)
+              return CreateBoundsUnknown();
```

Comment:
This is `bounds(none)` for the moment, but we will eventually want it to be the bounds if both the LHS and RHS have the same bounds. I almost wanted it to be `bounds(any)` but that seems too wide.